File: LTE_record_process/+ltepipe/Pipeline.m
classdef Pipeline < handle
%PIPELINE Connect uniformly registered streaming modules.

    properties (SetAccess = private)
        ExecutionConfig
        Runtime
        Modules = cell(1, 0)
        ModuleNames = cell(1, 0)
        State = 'created'
        Initialized = false
        Iterations = 0
        Finalized = false
        TerminationReason = ''
        FinalArtifacts = cell(1, 0)
        LastMessage
    end

    methods
        function obj = Pipeline(executionConfig)
            if nargin < 1 || isempty(executionConfig)
                executionConfig = struct( ...
                    'MaximumIterations', Inf, 'Verbose', false);
            end
            obj.ExecutionConfig = executionConfig;
            obj.Runtime = ltepipe.Runtime();
            obj.LastMessage = ltepipe.Message.none();
        end

        function register(obj, module)
            if obj.Initialized
                error('ltepipe:Pipeline:AlreadyInitialized', ...
                    'Modules cannot be registered after initialization.');
            end
            if ~isa(module, 'ltepipe.Module')
                error('ltepipe:Pipeline:InvalidModule', ...
                    'Registered objects must inherit from ltepipe.Module.');
            end
            if any(strcmp(obj.ModuleNames, module.Name))
                error('ltepipe:Pipeline:DuplicateModuleName', ...
                    'Module name "%s" is already registered.', module.Name);
            end
            obj.validateConnection(module);
            obj.Modules{end+1} = module;
            obj.ModuleNames{end+1} = module.Name;
        end

        function initialize(obj)
            if obj.Initialized
                return;
            end
            if isempty(obj.Modules)
                error('ltepipe:Pipeline:NoModules', ...
                    'At least one module must be registered.');
            end
            if ~isempty(obj.Modules{1}.InputType)
                error('ltepipe:Pipeline:MissingSource', ...
                    'The first registered module must be a source.');
            end
            obj.State = 'initializing';
            for index = 1:numel(obj.Modules)
                obj.Modules{index}.initialize(obj.Runtime);
            end
            obj.Initialized = true;
            obj.State = 'ready';
        end

        function result = step(obj)
            if obj.Finalized
                error('ltepipe:Pipeline:AlreadyFinalized', ...
                    'The pipeline cannot process data after finalization.');
            end
            obj.initialize();
            obj.State = 'running';
            obj.Iterations = obj.Iterations+1;
            result = obj.runModules(1, ltepipe.Message.none());
            obj.LastMessage = result.Message;
            if strcmp(result.Directive, 'stop')
                obj.State = 'stopping';
            else
                obj.State = 'ready';
            end
        end

        function summary = run(obj)
            maximum = obj.maximumIterations();
            terminationReason = 'maximum-iterations';
            try
                while obj.Iterations < maximum
                    result = obj.step();
                    if strcmp(result.Directive, 'stop')
                        terminationReason = result.Reason;
                        break;
                    end
                end
                obj.finalize(terminationReason);
            catch processingError
                obj.State = 'failed';
                try
                    obj.finalize('failed');
                catch finalizationError
                    warning('ltepipe:Pipeline:FinalizationFailed', ...
                        'Finalization after failure also failed: %s', ...
                        finalizationError.message);
                end
                rethrow(processingError);
            end
            summary = obj.getStatus();
            if obj.verbose()
                fprintf('Pipeline completed after %d iterations (%s).\n', ...
                    obj.Iterations, obj.TerminationReason);
            end
        end

        function artifacts = finalize(obj, reason)
            if nargin < 2 || isempty(reason)
                reason = 'completed';
            end
            if obj.Finalized
                artifacts = obj.FinalArtifacts;
                return;
            end
            obj.initialize();
            obj.State = 'finalizing';
            artifacts = cell(1, 0);
            for index = 1:numel(obj.Modules)
                moduleResult = obj.Modules{index}.finalize(reason);
                ltepipe.Result.validate(moduleResult);
                obj.dispatchCommands(moduleResult.Commands);
                if ltepipe.Message.hasContent(moduleResult.Message)
                    routedResult = obj.runModules( ...
                        index+1, moduleResult.Message);
                    artifacts = [artifacts, ...
                        routedResult.Message.Artifacts]; %#ok<AGROW>
                end
            end
            obj.FinalArtifacts = artifacts;
            obj.TerminationReason = char(reason);
            obj.Finalized = true;
            if strcmp(reason, 'failed')
                obj.State = 'failed';
            else
                obj.State = 'finalized';
            end
        end

        function module = getModule(obj, name)
            index = find(strcmp(obj.ModuleNames, char(name)), 1);
            if isempty(index)
                error('ltepipe:Pipeline:UnknownModule', ...
                    'Module "%s" is not registered.', char(name));
            end
            module = obj.Modules{index};
        end

        function status = getStatus(obj)
            moduleCount = numel(obj.Modules);
            moduleStatus = repmat(struct( ...
                'Name', '', 'Class', '', 'Status', struct()), ...
                1, moduleCount);
            for index = 1:moduleCount
                moduleStatus(index).Name = obj.ModuleNames{index};
                moduleStatus(index).Class = class(obj.Modules{index});
                moduleStatus(index).Status = ...
                    obj.Modules{index}.getStatus();
            end
            status = struct( ...
                'State', obj.State, ...
                'Iterations', obj.Iterations, ...
                'Initialized', obj.Initialized, ...
                'Finalized', obj.Finalized, ...
                'TerminationReason', obj.TerminationReason, ...
                'Modules', moduleStatus, ...
                'Runtime', obj.Runtime.getStatus(), ...
                'FinalArtifacts', {obj.FinalArtifacts});
        end
    end

    methods (Access = private)
        function result = runModules(obj, firstIndex, message)
            result = ltepipe.Result.forward(message);
            for index = firstIndex:numel(obj.Modules)
                result = obj.Modules{index}.process(result.Message);
                ltepipe.Result.validate(result);
                obj.dispatchCommands(result.Commands);
                switch result.Directive
                    case 'continue'
                        continue;
                    case 'reset-downstream'
                        obj.resetModulesAfter(index, result);
                        return;
                    case 'stop'
                        return;
                end
            end
        end

        function resetModulesAfter(obj, sourceIndex, result)
            event = struct( ...
                'Epoch', result.Epoch, ...
                'Reason', result.Reason, ...
                'Source', obj.ModuleNames{sourceIndex});
            for index = sourceIndex+1:numel(obj.Modules)
                obj.Modules{index}.reset(event);
            end
        end

        function dispatchCommands(obj, commands)
            for index = 1:numel(commands)
                command = commands{index};
                if ~isstruct(command) || ...
                        ~all(isfield(command, {'Target', 'Type'}))
                    error('ltepipe:Pipeline:InvalidCommand', ...
                        'Commands require Target and Type fields.');
                end
                target = obj.getModule(command.Target);
                target.handleCommand(command);
            end
        end

        function validateConnection(obj, module)
            if isempty(obj.Modules)
                return;
            end
            previous = obj.Modules{end};
            if ~isempty(previous.OutputType) && ...
                    ~isempty(module.InputType) && ...
                    ~strcmp(previous.OutputType, module.InputType)
                error('ltepipe:Pipeline:IncompatibleModules', ...
                    ['Module "%s" outputs "%s", but module "%s" ' ...
                    'expects "%s".'], previous.Name, ...
                    previous.OutputType, module.Name, module.InputType);
            end
        end

        function maximum = maximumIterations(obj)
            if isfield(obj.ExecutionConfig, 'MaximumIterations')
                maximum = obj.ExecutionConfig.MaximumIterations;
            elseif isfield(obj.ExecutionConfig, 'MaximumSubframes')
                maximum = obj.ExecutionConfig.MaximumSubframes;
            else
                maximum = Inf;
            end
            validateattributes(maximum, {'numeric'}, ...
                {'scalar', 'positive'});
        end

        function value = verbose(obj)
            value = isfield(obj.ExecutionConfig, 'Verbose') && ...
                obj.ExecutionConfig.Verbose;
        end
    end
end

File: LTESen_python/ltesen/ltepipe/pipeline.py
"""Generic linear module pipeline."""

from __future__ import annotations

import math
from typing import Any

from .message import Message
from .module import Module
from .result import Result, validate_result
from .runtime import Runtime


class Pipeline:
    def __init__(self, execution_config: dict[str, Any] | None = None) -> None:
        self.execution_config = execution_config or {
            "maximum_iterations": math.inf,
            "verbose": False,
        }
        self.runtime = Runtime()
        self.modules: list[Module] = []
        self.state = "created"
        self.initialized = False
        self.iterations = 0
        self.finalized = False
        self.termination_reason = ""
        self.final_artifacts: list[Any] = []
        self.last_message = Message.none()

    @property
    def module_names(self) -> list[str]:
        return [module.name for module in self.modules]

    def register(self, module: Module) -> None:
        if self.initialized:
            raise RuntimeError("Modules cannot be registered after initialization")
        if not isinstance(module, Module):
            raise TypeError("Registered objects must be Module instances")
        if module.name in self.module_names:
            raise ValueError(f'Module name "{module.name}" is already registered')
        self._validate_connection(module)
        self.modules.append(module)

    def initialize(self) -> None:
        if self.initialized:
            return
        if not self.modules:
            raise RuntimeError("At least one module must be registered")
        if self.modules[0].input_type:
            raise RuntimeError("The first registered module must be a source")
        self.state = "initializing"
        for module in self.modules:
            module.initialize(self.runtime)
        self.initialized = True
        self.state = "ready"

    def step(self) -> Result:
        if self.finalized:
            raise RuntimeError("The pipeline cannot process data after finalization")
        self.initialize()
        self.state = "running"
        self.iterations += 1
        result = self._run_modules(0, Message.none())
        self.last_message = result.message
        self.state = "stopping" if result.directive == "stop" else "ready"
        return result

    def run(self) -> dict[str, Any]:
        maximum = self._maximum_iterations()
        termination_reason = "maximum-iterations"
        try:
            while self.iterations < maximum:
                result = self.step()
                if result.directive == "stop":
                    termination_reason = result.reason
                    break
            self.finalize(termination_reason)
        except Exception:
            self.state = "failed"
            try:
                self.finalize("failed")
            except Exception:
                pass
            raise

        summary = self.get_status()
        if self._verbose():
            print(
                f"Pipeline completed after {self.iterations} iterations "
                f"({self.termination_reason})."
            )
        return summary

    def finalize(self, reason: str = "completed") -> list[Any]:
        if self.finalized:
            return self.final_artifacts
        self.initialize()
        self.state = "finalizing"
        artifacts: list[Any] = []
        for index, module in enumerate(self.modules):
            module_result = module.finalize(reason)
            validate_result(module_result)
            self._dispatch_commands(module_result.commands)
            if module_result.message.has_content:
                routed = self._run_modules(index + 1, module_result.message)
                artifacts.extend(routed.message.artifacts)
        self.final_artifacts = artifacts
        self.termination_reason = reason
        self.finalized = True
        self.state = "failed" if reason == "failed" else "finalized"
        return artifacts

    def get_module(self, name: str) -> Module:
        for module in self.modules:
            if module.name == name:
                return module
        raise KeyError(f'Module "{name}" is not registered')

    def get_status(self) -> dict[str, Any]:
        return {
            "state": self.state,
            "iterations": self.iterations,
            "initialized": self.initialized,
            "finalized": self.finalized,
            "termination_reason": self.termination_reason,
            "modules": [
                {
                    "name": module.name,
                    "class": type(module).__name__,
                    "status": module.get_status(),
                }
                for module in self.modules
            ],
            "runtime": self.runtime.get_status(),
            "final_artifacts": self.final_artifacts,
        }

    def _run_modules(self, first_index: int, message: Message) -> Result:
        result = Result.forward(message)
        for index in range(first_index, len(self.modules)):
            result = self.modules[index].process(result.message)
            validate_result(result)
            self._dispatch_commands(result.commands)
            if result.directive == "continue":
                continue
            if result.directive == "reset-downstream":
                event = {
                    "epoch": result.epoch,
                    "reason": result.reason,
                    "source": self.modules[index].name,
                }
                for module in self.modules[index + 1 :]:
                    module.reset(event)
            return result
        return result

    def _dispatch_commands(self, commands: tuple[Any, ...]) -> None:
        for command in commands:
            if not isinstance(command, dict) or not all(
                key in command for key in ("target", "type")
            ):
                if isinstance(command, dict) and all(
                    key in command for key in ("Target", "Type")
                ):
                    command = {
                        "target": command["Target"],
                        "type": command["Type"],
                        **command,
                    }
                else:
                    raise ValueError("Commands require target and type fields")
            self.get_module(command["target"]).handle_command(command)

    def _validate_connection(self, module: Module) -> None:
        if not self.modules:
            return
        previous = self.modules[-1]
        if previous.output_type and module.input_type and previous.output_type != module.input_type:
            raise ValueError(
                f'Module "{previous.name}" outputs "{previous.output_type}", '
                f'but module "{module.name}" expects "{module.input_type}"'
            )

    def _maximum_iterations(self) -> float:
        value = self.execution_config.get(
            "maximum_iterations", self.execution_config.get("maximum_subframes", math.inf)
        )
        if value is None:
            return math.inf
        if not isinstance(value, (int, float)) or value <= 0:
            raise ValueError("maximum_iterations must be a positive number or null")
        return float(value)

    def _verbose(self) -> bool:
        return bool(self.execution_config.get("verbose", False))


__all__ = ["Pipeline"]
